tools/crm_node.c

```diff
@@ -540,49 +573,41 @@ static int
 purge_node_from_cib(const char *node_name, long node_id)
 {
     int rc = pcmk_rc_ok;
+    int commit_rc = pcmk_rc_ok;
     cib_t *cib = NULL;
-    xmlNode *xml = NULL;
 
-    // Connect to CIB
+    // Connect to CIB and start a transaction
     cib = cib_new();
     if (cib == NULL) {
         return ENOTCONN;
     }
     rc = cib->cmds->signon(cib, crm_system_name, cib_command);
-    if (rc != pcmk_ok) {
-        rc = pcmk_legacy2rc(rc);
-        goto done;
+    if (rc == pcmk_ok) {
+        rc = cib->cmds->init_transaction(cib);
     }
-
-    // Remove from configuration section
-    xml = create_xml_node(NULL, XML_CIB_TAG_NODE);
-    crm_xml_add(xml, XML_ATTR_UNAME, node_name);
-    if (node_id > 0) {
-        crm_xml_set_id(xml, "%ld", node_id);
-    }
-    rc = cib->cmds->remove(cib, XML_CIB_TAG_NODES, xml, cib_sync_call);
     if (rc != pcmk_ok) {
         rc = pcmk_legacy2rc(rc);
-        goto done;
+        cib__clean_up_connection(&cib);
+        return rc;
     }
-    free_xml(xml);
 
-    // Remove from status section
-    xml = create_xml_node(NULL, XML_CIB_TAG_STATE);
-    crm_xml_add(xml, XML_ATTR_UNAME, node_name);
-    if (node_id > 0) {
-        crm_xml_set_id(xml, "%ld", node_id);
-    }
-    rc = cib->cmds->remove(cib, XML_CIB_TAG_STATUS, xml, cib_sync_call);
-    rc = pcmk_legacy2rc(rc);
+    // Remove from configuration and status
+    rc = remove_from_section(cib, XML_CIB_TAG_NODE, XML_CIB_TAG_NODES,
+                             node_name, node_id);
     if (rc == pcmk_rc_ok) {
-        crm_debug("Purged node %s (%ld) from CIB",
-                  pcmk__s(node_name, "by ID"), node_id);
+        rc = remove_from_section(cib, XML_CIB_TAG_STATE, XML_CIB_TAG_STATUS,
+                                 node_name, node_id);
     }
 
-done:
-    free_xml(xml);
+    // Commit the transaction
+    commit_rc = cib->cmds->end_transaction(cib, (rc == pcmk_rc_ok),
+                                           cib_sync_call);
     cib__clean_up_connection(&cib);
+
+    if ((rc == pcmk_rc_ok) && (commit_rc == pcmk_ok)) {
+        crm_debug("Purged node %s (%ld) from CIB",
```

Comment:
Error message on failure too? If it's a server-side failure, there might already be something in the daemon logs but it would go unreported to the client.

Reply:
The caller will